revocation/revocation.go

```diff
@@ -0,0 +1,249 @@
+		return nil, fmt.Errorf("failed to retrieve OSCP: response had status code %d", resp.StatusCode)
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case bytes.Equal(body, ocsp.UnauthorizedErrorResponse):
+		return nil, errors.New("OSCP unauthorized")
+	case bytes.Equal(body, ocsp.MalformedRequestErrorResponse):
+		return nil, errors.New("OSCP malformed")
+	case bytes.Equal(body, ocsp.InternalErrorErrorResponse):
+		return nil, errors.New("OSCP internal error")
+	case bytes.Equal(body, ocsp.TryLaterErrorResponse):
+		return nil, errors.New("OSCP try later")
+	case bytes.Equal(body, ocsp.SigRequredErrorResponse):
+		return nil, errors.New("OSCP signature required")
+	}
+
+	return ocsp.ParseResponseForCert(body, cert, issuer)
+}
+
+// SetMergeErrorsFunction allows you to specify an alternative function to merge errors if the default does not fit your use case. You can also pass nil to reset it to the defaultMergeErrors function
+func (r *revocation) SetMergeErrorsFunction(mergeErrorsFunc func(errors []error) error) {
+	if mergeErrorsFunc == nil {
+		r.mergeErrorsFunc = defaultMergeErrors
+	} else {
+		r.mergeErrorsFunc = mergeErrorsFunc
+	}
+}
+
+// DefaultMergeErrors condenses errors for a list of errors (either for cert chain or OCSP servers) into one primary error
+func defaultMergeErrors(errorList []error) error {
+	var result error
+	if len(errorList) > 0 {
+		result = errorList[0]
+
+		for _, err := range errorList {
+			if err == nil {
+				continue
+			}
+			switch t := err.(type) {
+			case RevokedInOCSPError:
+				// There is a revoked certificate
+				// return since any cert being revoked means leaf is revoked
+				return t
+			case CheckOCSPError:
+				// There is an error checking
+				// return since any cert having error means chain has error (return earliest)
+				return t
+			case UnknownInOCSPError:
+				// A cert in the chain has status unknown
+				// will not return immediately (in case one is revoked or has error), but will override other chain errors
+				result = t
+			case NoOCSPServerError:
+				// A cert in the chain does not have OCSP enabled
+				// Still considered valid and not revoked
+				// will not return immediately (in case there is higher level error)
+				// will override OCSPTimeoutError and nil, but not UnknownInOCSPError (since a known unknown is worse than a cert without OCSP)
+				if _, ok := result.(UnknownInOCSPError); !ok || result == nil {
+					result = t
+				}
+			case OCSPTimeoutError:
+				// A cert in the chain timed out while checking OCSP
+				// will not return immediately (in case there is higher level error)
+				// will override nil, but not UnknownInOCSPError or NoOCSPServerError (since timeout should only be conveyed if that is the only issue)
+				if result == nil {
+					result = t
+				}
+			default:
+				return CheckOCSPError{Err: err}
+			}
+		}
+
+		return result
+	} else {
+		return nil
+	}
```

Comment:
How are we deciding(why above priority?) what singular error to through?

Reply:
I think there are 2 parts to this: 1. why this priority, and 2. why we are condensing a list of errors to 1 error

---

I chose this priority based on how close the error is to an actual revocation (a known negative)
 - If any of the certs have a RevokedInOCSPError, then it makes sense for the chain to also have RevokedInOCSPError.
 - If there is an unavoidable error with the chain, then that issue needs to be resolved before revocation can be determined, hence the first of these errors is returned.
 - If a cert has UnknownInOCSPError, then it isn't necessarily revoked. We should go through the rest of the chain to see if any are actually revoked or had errors (known negative), otherwise we should say that the status is unknown. Since this is a status from the server (which should be able to return a status of good), it should not be considered positive (hence its increased priority)
 - If a cert has NoOCSPServerError, then its OCSP status cannot ever be determined. It should never be considered revoked by OCSP, but shouldn't necessarily be considered valid either. It is an unknown, but will always be unknown. The rest of the chain should be checked without this error inhibiting them.
 - If there is an OCSPTimeoutError, then the status is temporarily unknown. It should be the least pressing, since it can be resolved by simply updating the timeout for the client (unless a connection cannot be made). If there are other issues that need to be resolved before this (i.e. the ones above) then it will not take priority since it is a temporary unknown.

Knowing that this prioritization might not be ideal for every use case, I also added the ability to change how errors are condensed via the `SetMergeErrorsFunction` method, allowing users to choose their own solution for how to condense the errors if the default does not match their interpretation.

---

I agree that having a priority to condense errors to a singular response isn't the best, but I think it makes sense to return a singular error to the user. If we are validating an entire chain for OCSP (or CRL for that matter), we would potentially have a different error for each certificate in the chain. Then, for each certificate, there could be multiple OCSP servers that each have a unique error. It just doesn't make sense to return all of these errors to the user and force them to parse through it to check if a chain is revoked.

An alternative might be to still condense for OCSP server errors (so that there is one error for each cert) then return a list of errors for the entire chain. Then, we could provide a function that converts this list into a single error for ease of use (i.e. not forcing this conversion, but still allowing it)

Reply:
Resolving, further discussion tracked [here](https://github.com/notaryproject/notation-core-go/pull/134#discussion_r1151422211)